doc/codegen/modular_codegen_design.md

```diff
@@ -0,0 +1,561 @@
+│   ├── transformation/     # Transformation passes
+│   │   ├── __init__.py
+│   │   ├── copy_to_map.py
+│   │   ├── stream_assignment.py
+│   │   ├── tasklet_lowering.py
+│   │   └── memlet_lowering.py
+│   ├── codegen/           # Code generation passes
+│   │   ├── __init__.py
+│   │   ├── state_struct.py
+│   │   ├── allocation_code.py
+│   │   ├── frame_code.py
+│   │   ├── target_code.py
+│   │   └── header_generation.py
+│   ├── file_generation/   # File organization passes
+│   │   ├── __init__.py
+│   │   ├── sdfg_splitting.py
+│   │   └── code_objects.py
+│   └── pipelines.py       # Pre-built pipelines
+├── targets/               # Target-specific generators (simplified)
+│   ├── __init__.py
+│   ├── base.py            # Base target interface (from target.py)
+│   ├── openmp.py          # OpenMP backend (split from cpu.py)
+│   ├── cpp.py             # Pure C++ backend
+│   ├── gpu.py             # GPU backend (generalized from cuda.py)
+│   ├── cuda.py            # CUDA-specific GPU
+│   ├── hip.py             # HIP-specific GPU
+│   ├── fpga/              # FPGA backends
+│   └── specialized/       # Other specialized targets
+├── runtime/               # Runtime interface (from compiled_sdfg.py)
+└── utils/                 # Utilities (dispatcher, codeobject, etc.)
+    ├── __init__.py
+    ├── dispatcher.py
+    ├── codeobject.py
+    ├── prettycode.py
+    └── common.py
+```
+
+## Target Refactoring
+
+### Current Issues
+- **CPU backend** actually does OpenMP generation
+- **CUDA backend** is GPU-specific, not general GPU
+- Poor separation between generic and specific backends
+
+### Proposed Refactoring
+
+#### 1. **C++ Backend** (`targets/cpp.py`)
+- Pure C++ code generation without parallelization
+- Base for other C++ based backends
+- Sequential execution model
+- Basic memory management
+
+#### 2. **OpenMP Backend** (`targets/openmp.py`)
+- Extends C++ backend with OpenMP directives
+- CPU parallelization via OpenMP
+- Current "CPU" backend functionality
+- Shared memory parallelism
+
+#### 3. **GPU Backend** (`targets/gpu.py`)
+- Generic GPU programming model
+- Common GPU concepts (kernels, memory hierarchy)
+- Base for CUDA/HIP/OpenCL backends
+- Device memory management
+
+#### 4. **CUDA Backend** (`targets/cuda.py`)
+- Extends GPU backend with CUDA specifics
+- CUDA runtime API calls
+- CUDA-specific optimizations
+- PTX inline assembly support
+
+#### 5. **HIP Backend** (`targets/hip.py`)
+- Extends GPU backend with HIP specifics
+- AMD GPU support
+- ROCm integration
+- Portable GPU code
+
+### Target Hierarchy
+```
+TargetCodeGenerator (base)
+├── CppCodeGen (sequential C++)
+│   ├── OpenMPCodeGen (CPU parallelism)
+│   └── MPICodeGen (distributed)
+├── GPUCodeGen (generic GPU)
+│   ├── CUDACodeGen (NVIDIA)
+│   ├── HIPCodeGen (AMD)
+│   └── OpenCLCodeGen (portable)
+├── FPGACodeGen (FPGA base)
+│   ├── XilinxCodeGen
+│   └── IntelFPGACodeGen
+└── SpecializedCodeGen
+    ├── SVECodeGen (ARM vectors)
```

Comment:
SVE code generation is not really a specialized code generation, but rather a scope generator just like OpenMP.